CARDS-2288: New question type: Selectable Areas

acrowthe opened the pull request:
Add new SelectableAreaQuestion
- This question type supports multiple `variants`
  - Each variant comes with an image and a list of different selectable areas on the image
  - Selectable areas can be defined as either rectangles or polygons with configurable colors
- Each selectable area can have configurable unselected fill color, selected fill color and stroke color
  - Color configuration can be done on an individual area, variant or question level
  - Selected color defaults to an opaque version of the primary theme color
- Each question determines what areas from the selected variant are available using AnswerOptions
  - If a `notApplicable` AnswerOption is present, a checkbox will be presented above the image for this option
- The list of current selected areas is displayed beside the image or below if the screen is narrow
- Hovering areas while in edit mode will highlight the current hovered area and display a tooltip with it's label
- In view mode, if `notApplicable` is selected the image will not be displayed
  - The image will otherwise be displayed with hover highlights disabled and tooltips only enabled on selected areas
- Initially 1 variant is provided for selectable areas of the body
  - Configuration is set up so projects can add their own variants with custom images and maps

- Configuration for hover colors is present in the question and variant specification but is not currently supported by the front end code. A fixed hover color based on an opaque varient of the primary text color is used

A test questionnaire is provided (SelectableAreasTest)

acrowthe committed:
CARDS-2288: New question type: Selectable Areas
Add new SelectableAreaQuestion
- This question type supports multiple `variants`
  - Each variant comes with an image and a list of different selectable areas on the image
  - Selectable areas can be defined as either rectangles or polygons with configurable colors
- Each selectable area can have configurable unselected fill color, selected fill color and stroke color
  - Color configuration can be done on an individual area, variant or question level
  - Selected color defaults to an opaque version of the primary theme color
- Each question determines what areas from the selected variant are available using AnswerOptions
  - If a `notApplicable` AnswerOption is present, a checkbox will be presented above the image for this option
- The list of current selected areas is displayed beside the image or below if the screen is narrow
- Hovering areas while in edit mode will highlight the current hovered area and display a tooltip with it's label
- In view mode, if `notApplicable` is selected the image will not be displayed
  - The image will otherwise be displayed with hover highlights disabled and tooltips only enabled on selected areas
- Initially 1 variant is provided for selectable areas of the body
  - Configuration is set up so projects can add their own variants with custom images and maps

- Configuration for hover colors is present in the question and variant specification but is not currently supported
  by the front end code. A fixed hover color based on an opaque varient of the primary text color is used

A test questionnaire is provided (SelectableAreasTest)

acrowthe commented:
Bug: WSOD when using the questionnaire designer to edit a form with Selectable Area questions. Bug carried over from previous 2288 implementation (https://github.com/data-team-uhn/cards/pull/1654)

acrowthe commented:
https://github.com/data-team-uhn/cards/assets/20056751/9eed9d4f-6f3e-4071-ab5c-0a5cff299351

acrowthe committed:
CARDS-2288: New question type: Selectable Areas
- Fix console error due to missing key
- Remove variant from questionnaire designer preventing WSOD

acrowthe commented:
Questionnaire designer WSOD somewhat fixed. it will no longer crash but there is no way to enter the variant for a selectableArea question so the questionnaire designer part of selectableAreas still needs to be improved

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -705,6 +706,17 @@ const questionnaireStyle = theme => ({
+      "& rect:hover": {
+        fill: alpha(theme.palette.text.primary, 0.4)
+      }
+    }
```
let's make the widget self-contained and move the styling to SelectableAreaQuestion.jsx

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SelectableAreaQuestion.jsx:
```diff
@@ -0,0 +1,401 @@
+  let getUnselectedColor = (area) => getColor(area, "unselectedColor", "transparent")
+  let getSelectedColor = (area) => getColor(area, "selectedColor", theme.palette.primary.main + "66")
+  // TODO: reimplement hover color configuration. Currently hover color is fixed by QuestionnaireStyle
+  let getHoverColor = (area) => getColor(area, "hoverColor", alpha(theme.palette.text.primary, 0.4));
```
Hover color should depend on whether the area is selected. We should experiment with these to see which work best (note they already have transparency): 
![image](https://github.com/data-team-uhn/cards/assets/651980/b428c373-bc73-4bf5-b637-3bbc5a8cb69f)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SelectableAreaQuestion.jsx:
```diff
@@ -0,0 +1,401 @@
+  let getSelectedColor = (area) => getColor(area, "selectedColor", theme.palette.primary.main + "66")
+  // TODO: reimplement hover color configuration. Currently hover color is fixed by QuestionnaireStyle
+  let getHoverColor = (area) => getColor(area, "hoverColor", alpha(theme.palette.text.primary, 0.4));
+  let getStrokeColor = (area) => getColor(area, "strokeColor", theme.palette.text.primary);
```
I would use `theme.palette.divider` here by default.

acrowthe committed:
CARDS-2288: New question type: Selectable Areas
- Change single select areas so they deselect when clicked on again

acrowthe commented:
Rebased onto dev to pull in compact sections.
Changed single select selectable areas to toggle selection when the same area is selected repeatedly

marta- committed:
CARDS-2288: New question type: Selectable Areas
Adapted the ListInput from questionnaire designer to single-select, enable `variant` selection for Selectable Areas questions

marta- commented:
There's a WSORD when displaying a selectable areas question that doesn't have a valid `variant` specified. I would suggest logging a console error, but handling more gracefully in the UI by falling back to a plain multiple choice.

marta- commented:
The questionnaire designer was slightly updated in commit de63fc7 to support `variant` selection for selectable areas questions. We will need **regression testing** for subject selection and vocabulary selection @sashaandjic .

sashaandjic commented:
The second page of the Clinician Assessment does not get generated. 
Also, I think it should be Left and Right buttock (not buttocks).
![image](https://github.com/data-team-uhn/cards/assets/23243244/59535674-4040-496a-9e4c-1cf3b19a9cc4)

acrowthe committed:
CARDS-2288: New question type: Selectable Areas
Reimplement configurable hover color and improve color asasignment code

acrowthe commented:
Hover styling has been improved and is now darker when hovering a selected area in addition to being configurable:

https://github.com/data-team-uhn/cards/assets/20056751/5bdd58e7-4ace-4a14-971b-bd1b4261bc97

Clinician assessment page 2 not showing up is an XML error unrelated to this PR

acrowthe committed:
CARDS-2288: New question type: Selectable Areas
Convert body image to svg

acrowthe committed:
CARDS-2288: New question type: Selectable Areas
Begin adding new variant for selectable areas of the face
- Base variant is functional, will need additional areas added to the map based on required options

acrowthe commented:
@sashaandjic looks like the original PDF uses the plural for buttocks even in the singular case:

![image](https://github.com/data-team-uhn/cards/assets/20056751/3824a6b8-9f60-404c-8fe1-975b7a618db8)

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -18,6 +18,7 @@
 //
 
 import { orange, grey } from '@mui/material/colors';
+import { alpha } from '@mui/material/styles';
```
This import doesn't seem used in here, can it be removed?

sdumitriu commented on modules/data-entry/src/main/media/SLING-INF/content/libs/cards/resources/media/dataEntry/fullBody.svg:
```diff
@@ -0,0 +1,62 @@
+       width="1280"
+       height="1280"
+       preserveAspectRatio="none"
+       xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAABQAAAAUACAIAAACXhmigAAABhGlDQ1BJQ0MgcHJvZmlsZQAAeJx9&#10;kT1Iw0AcxV9TRdGKg0WkOASpTnZREcdSxSJYKG2FVh1MLv2CJg1Jiouj4Fpw8GOx6uDirKuDqyAI&#10;foC4ujgpukiJ/0sKLWI8OO7Hu3uPu3eA0Kgw1eyKAqpmGal4TMzmVsWeV/RjGEGMISQxU0+kFzPw&#10;HF/38PH1LsKzvM/9OQaUvMkAn0gcZbphEW8Qz25aOud94iArSQrxOfGkQRckfuS67PIb56LDAs8M&#10;GpnUPHGQWCx2sNzBrGSoxDPEYUXVKF/Iuq ...
```
I think this can be deleted.

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/content/libs/cards/dataEntry/SelectableArea/Face.json:
```diff
@@ -0,0 +1,12 @@
+    "imageWidth": 1280,
+    "imageHeight": 1280,
+    "strokeColor": "#00000000",
+    "hoverColor": "#55555555",
```
I think these should be default values that don't need to be specified.

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -95,4 +95,40 @@
+
+//-----------------------------------------------------------------------------
+// The configuration for a selectable area question variant
+[cards:SelectableAreaVariant] > sling:Folder, mix:referenceable
```
I feel that right now a better place for this would be in `modules/data-model/forms/api/src/main/resources/SLING-INF/nodetypes/forms.cnd`.

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -95,4 +95,40 @@
+  // Hardcode the resource type.
+  - sling:resourceType (STRING) = "cards/SelectableAreaVariant" mandatory autocreated protected
+
+  // Hardcode the resource supertype: the  is a resource homepage.
```
Update.

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -95,4 +95,40 @@
+  // The image that should be displayed for this variant
+  - image (STRING) mandatory
+
+  // The data mapp that should be applied to this variant
```
```suggestion
  // The data map that should be applied to this variant
```

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -95,4 +95,40 @@
+
+  // The color that the areas should be highlighted in when selected.
+  // If present, `fillColor` as specified in the image map will be used as a hover color
+  // and this color will be usedfor selected areas.
```
```suggestion
  // and this color will be used for selected areas.
```

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -95,4 +95,40 @@
+  // The color that the areas should be highlighted in when selected.
+  // If present, `fillColor` as specified in the image map will be used as a hover color
+  // and this color will be usedfor selected areas.
+  - highlightColor (STRING)
```
I also see `imageHeight`, `imageWidth`, `hoverColor` used in the actual nodes, but not defined in the cnd.

sdumitriu replied:
I see a lot of other properties used in the frontend code, so maybe just add a wildcard property definition at the end.

sdumitriu commented on modules/data-model/forms/api/src/main/resources/SLING-INF/content/oak%3Aindex/forms.json:
```diff
@@ -428,6 +428,33 @@
+                    "propertyIndex": true,
+                    "notNullCheckEnabled": true,
+                    "type": "String"
+                },
```
Also add the `copiedFrom` definition from the other answers.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SelectableAreaQuestion.jsx:
```diff
@@ -0,0 +1,412 @@
+import AnswerComponentManager from "./AnswerComponentManager";
+import FormattedText from "../components/FormattedText.jsx";
+
+// Component that renders a multiple choice question, with optional text input.
```
I feel that this is underselling it. Mention that the question is displayed as a clickable image with different areas as the options to choose from.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SelectableAreaQuestion.jsx:
```diff
@@ -0,0 +1,412 @@
+  }
+
+  let updateAreaColor = (area) => {
+    area.fill = (area.isHovered ? getHoverColor(area)
+      : (isAreaSelected(area) ? getSelectedColor(area) : getUnselectedColor(area)));
```
I would also like a hover+selected color.

sdumitriu commented on modules/data-model/forms/api/src/main/resources/SLING-INF/nodetypes/forms.cnd:
```diff
@@ -154,6 +154,7 @@
   // - pedigree: a pedigree drawing
   // - phone: a phone or fax number
   // - address: a postal address
+  // - selectableArea: a area selected from an image
```
Is this a data type, or a display mode? I feel that it's more of a display mode.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SelectableAreaQuestion.jsx:
```diff
@@ -0,0 +1,412 @@
+    if (map) {
+      setMap(
+        map.map((mapEntry => {
+          let entry = mapEntry
```
```suggestion
          let entry = mapEntry;
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SelectableAreaQuestion.jsx:
```diff
@@ -0,0 +1,412 @@
+    // This allows for questions to only provide a subset of areas supported by the variant.
+    let outputMap = [];
+    inputMap.forEach(mapEntry => {
+      if (mapEntry.value in options) {
```
Should it be possible to have a question with no defined answer options, in which case all the areas in the map will be used? Which also means that the map should define the area title as well, not just the internal value?

On one hand, that would avoid duplication, since right now we have to manually define a whole bunch of answer options that we already defined in the map. On the other hand, we do have overlapping areas that are probably incompatible to have all at once, e.g. jaw vs left/right jaw. Or we could make it so that areas aren't completely overlapping, make the left jaw and right jaw a bit smaller, so there is a bit of space available for clicking on the whole jaw.

Or we could already define more different maps, `detailedFullBodyMap`, `majorAreasFullBodyMap`, `detailedHeadMap`, `majorAreasHeadMap`, `detailedLeftArmMap`, `detailedRightHandMap`...

acrowthe replied:
I'll add in code to enable all areas if no options are specified. That way it is available if needed by future projects without core code changes.

Some of the areas are already relatively small, trying to fit everything in without overlapping doesn't feel like a great option in the body map case especially since the current use cases will result in unselectable gaps in the available areas.

Defining extra maps feels like even worse duplication than the current duplicated options since adjusting coordinates if something is off is quite tedious already.
Adding in some sort of sub-map that use the coordinates from an existing map (or can override or add them) but only expose a subset of the areas could help with avoiding duplication

acrowthe committed:
CARDS-2288: New question type: Selectable Areas
Code cleanup

sdumitriu reviewed (approved):
Code changes look good, didn't run it yet.

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/content/libs/cards/dataEntry/SelectableArea/Face.json:
```diff
@@ -0,0 +1,10 @@
+{
+    "jcr:primaryType": "cards:SelectableAreaVariant",
+    "name": "fullBody",
```
Wrong name.

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/content/libs/cards/dataEntry/SelectableArea/FullBody.json:
```diff
@@ -0,0 +1,9 @@
+{
+    "jcr:primaryType": "cards:SelectableAreaVariant",
+    "name": "fullBody",
```
Can this be pretty, like `Full body`?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question.json:
```diff
@@ -460,6 +461,28 @@
+          "type": "textOptions",
+          "childrenType": "cards:AnswerOption"
+        },
+        "unselectedColor": "string",
```
Idea for the future: add support for a "color" type, with a nice color picker.

sdumitriu committed:
CARDS-2288: New question type: Selectable Areas
Better names for the two default variants

sdumitriu reviewed (changes requested):
It's a bit awkward to define a new question in the editor, since all desired parts must be manually entered in the editor, copied from the map json. We should at least display a hint message pointing to that file.

Enabling notes for this type of question is buggy, notes are editable in view mode as well.

sdumitriu committed:
CARDS-2288: New question type: Selectable Areas
Bugfix: notes are editable in view mode

sdumitriu commented:
> Enabling notes for this type of question is buggy, notes are editable in view mode as well.

Fixed.

sdumitriu committed:
CARDS-2288: New question type: Selectable Areas
Add a hint in the questionnaire editor that answer options must be selected
